model/src/main/kotlin/utils/Extensions.kt

```diff
@@ -172,27 +172,52 @@ fun Identifier.getPurlType() =
  * [in the documentation](https://github.com/package-url/purl-spec/blob/master/README.rst#purl).
  * E.g. 'maven' for Gradle projects.
  */
-fun Identifier.toPurl() =
-    if (this == Identifier.EMPTY) {
-        ""
-    } else {
-        buildString {
-            append("pkg:")
-            append(getPurlType())
-
-            if (namespace.isNotEmpty()) {
-                append('/')
-                append(namespace.percentEncode())
-            }
+fun Identifier.toPurl() = if (this == Identifier.EMPTY) "" else createPurl(getPurlType(), namespace, name, version)
 
-            append('/')
-            append(name.percentEncode())
+/**
+ * Create the canonical [package URL](https://github.com/package-url/purl-spec) ("purl") based on given properties:
+ * [type] (which must be a String representation of a [PurlType] instance, [namespace], [name] and [version].
+ * Optional [qualifiers] may be given and will be appended to the purl as query parameters e.g.
+ * pkg:deb/debian/curl@7.50.3-1?arch=i386&distro=jessie
+ * Optional [subpath] may be given and will be appended to the purl e.g.
+ * pkg:golang/google.golang.org/genproto#googleapis/api/annotations
+ *
+ */
+fun createPurl(
+    type: String,
+    namespace: String,
+    name: String,
+    version: String,
+    qualifiers: Map<String, String> = emptyMap(),
+    subpath: List<String> = emptyList()
```

Comment:
I believe this should just be a single `String`, because this is not a list of paths, but a single subpath. Note that you should *not* `percentEncode()` the whole subpath string at once then, as you'd want to keep the `/`, but do something like (typed blindly)

```
if (subpath.isNotEmpty()) {
    val value = subpath.split('/').joinToString("/", prefix = "#") { it.percentEncode() }
    append(value)
}
```

Reply:
This is disputable as the specs says: `The subpath contains zero or more segments, separated by slash '/'`
https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst

We could have also renamed the parameter as `segments`.
However I am mildly opiniated on this so  I did the changes.